### PR TITLE
SNOW-1458122 Index name apis

### PR DESCRIPTION
**Pull request by sfc-gh-vbudati**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1458122, SNOW-1021837

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Implemented Index `name`, `names`, `rename`, and `set_names`. 

This behavior is slightly different from native pandas since not all references to an Index or a DataFrame/Series' index are kept track of - this would require a big redesign. Instead, an Index can keep track of its parent DataFrame/Series and modify them when necessary.

`_parent` keeps track of the parent Series/DataFrame and is set whenever `df.index` or `series.index` is called.

Related design doc: https://docs.google.com/document/d/1clzOiC0Wu9KtTf0MqFDWagPesvKF0rKEizfz5IZ43v4/edit

This PR also fixes bug SNOW-1021837 where the index column name was not being correctly updated after `reset_index`.

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 1]
```diff
@@ -132,7 +133,7 @@ def __init__(
 
         Parameters
         ----------
-        data : array-like (1-dimensional), pandas.Index, modin.pandas.Series, optional
+        data : array-like (1-dimensional), pandas.Index, modin.pandas.Series, SnowflakeQueryCompiler, optional
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:136]: this is a change made last week, qc will be given throw argument `query_compiler`.
```suggestion
        data : array-like (1-dimensional), pandas.Index, modin.pandas.Series, optional
```

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 2]
```diff
@@ -172,11 +173,13 @@ def __init__(
 
     def _init_index(
         self,
-        data: ArrayLike | native_pd.Index | Series | None,
+        data: ArrayLike | native_pd.Index | Series | DataFrame | None,
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:176]:
```suggestion
        data: ArrayLike | native_pd.Index | Series | None,
```
Can not create index from dataframe

**Review by sfc-gh-azhan — CHANGES_REQUESTED** [review]

File: tests/integ/modin/index/test_name.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        new_native_idx.names
+        == new_snow_idx.names
+        == ([None] if new_name is None else ["grade"])
+    )
+    assert native_df.index.name == snow_df.index.name == "score"
+
+
+@pytest.mark.parametrize("inplace", [True, False])
+@sql_count_checker(query_count=0)
+def test_index_rename_list(inplace):
```
Comment (sfc-gh-mvashishtha) [inline tests/integ/modin/index/test_name.py:245]: This doesn't seem to be working end-to-end with `to_pandas`. Should we fix or raise an error before the user calls `to_pandas`?

```python
i = pd.Index([1, 2, 3])
i.name = ['name1']
i.to_pandas()
```

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 4]
```diff
@@ -1545,8 +1571,29 @@ def rename(self) -> None:
...
+        --------
+        >>> idx = pd.Index(['A', 'C', 'A', 'B'], name='score')
+        >>> idx.rename('grade', inplace=False)
+        Index(['A', 'C', 'A', 'B'], dtype='object', name='grade')
+        >>> idx.rename('grade', inplace=True)
+
+        Note
+        ----
+        Native pandas only allows hashable types for names. Snowpark pandas allows
+        name to be any scalar or list-like type. If a type is used for the name,
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:1585]:
```suggestion
        name to be any scalar or list-like type. If a tuple is used for the name,
```

File: tests/integ/modin/index/test_name.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    # a MultiIndex.
+    native_idx = native_pd.Index(["A", "C", "A", "B"], name="score")
+    snow_idx = pd.Index(native_idx)
+    with pytest.raises(ValueError, match="Level must be None for non-MultiIndex"):
+        native_idx.set_names("grade", level=level)
+
+    # Verifying the results.
+    native_res = native_idx.set_names("grade")
+    snow_res = snow_idx.set_names("grade", level=level)
+    assert native_res.name == snow_res.name == "grade"
```
Comment (sfc-gh-mvashishtha) [inline tests/integ/modin/index/test_name.py:291]: Could you please also add a test case for SNOW-1021837 and mention it in the PR description? I think your change will fix that bug.

File: tests/integ/modin/index/test_name.py (hunk 3 above)
Comment (sfc-gh-vbudati) [reply tests/integ/modin/index/test_name.py:246]: I don't think it's valid to set the name like that. I added a check to the `name` property to not accept non-hashable types as a name. The discrepancy I'm pointing out here is that Index.set_names in pandas accepts a list of names but rename does not (documentation says that it should) - so I made list-like names work with both APIs.
However, the name itself cannot be a list.

**Review by sfc-gh-mvashishtha — APPROVED** [review]

**Review by sfc-gh-azhan — APPROVED** [review]

**Pull request merged**